File: src/main/kotlin/com/castab/resend/service/OutboxService.kt
package com.castab.resend.service

import com.castab.resend.data.acceptBatchMessages
import com.castab.resend.data.assignEntriesToBatch
import com.castab.resend.data.createOutboxBatch
import com.castab.resend.data.deleteOrphanOutboxEntries
import com.castab.resend.data.deleteOwnedBatch
import com.castab.resend.data.findBatchMessagesOrdered
import com.castab.resend.data.findOutboxBatch
import com.castab.resend.data.h
import com.castab.resend.data.isBatchLeaseOwned
import com.castab.resend.data.markBatchMessagesState
import com.castab.resend.data.reclaimBatch
import com.castab.resend.data.scheduleBatchRetry
import com.castab.resend.data.selectRetryableBatchId
import com.castab.resend.data.selectUnbatchedMessageIds
import com.castab.resend.data.tx
import com.castab.resend.domain.EmailMessage
import com.castab.resend.domain.EmailMessageState
import com.castab.resend.email.ResendApiError
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import java.time.OffsetDateTime
import java.time.ZoneOffset
import java.util.UUID

private const val OUTBOX_LEASE_MS = 2L * 60 * 1000
private const val PROVIDER_IDEMPOTENCY_SAFETY_MS = 23L * 60 * 60 * 1000
private val RETRY_DELAYS_MS = longArrayOf(60_000, 120_000, 300_000)

private data class ClaimedBatch(
    val id: String,
    val leaseToken: String,
    val attemptCount: Int,
    val firstAttemptAt: OffsetDateTime,
    val messages: List<EmailMessage>,
)

fun Services.drainEmailOutbox(limit: Int): JsonObject {
    val claimed = claimOutboxBatch(limit) ?: return emptyDrainResult()
    if (claimed.attemptCount > 1 && hasProviderWindowExpired(claimed)) {
        return finalizeIndeterminateBatch(claimed, "Outbox batch exceeded the provider idempotency window")
    }

    return try {
        val ids = configuredResend().sendBatch(
            claimed.messages.map(::buildSendEmailInput),
            "conversation-outbox/${claimed.id}",
        )
        if (ids.size != claimed.messages.size || ids.any { it.isNullOrEmpty() }) {
            scheduleBatchRetryResult(claimed, "invalid_batch_response")
        } else {
            finalizeAcceptedBatch(claimed, ids.map { it!! })
        }
    } catch (error: Throwable) {
        if (error is ResendApiError && error.status == 409 && error.code == "invalid_idempotent_request") {
            finalizeIndeterminateBatch(claimed, "Resend rejected a changed payload for the persisted batch key")
        } else if (isRetryableBatchError(error)) {
            scheduleBatchRetryResult(claimed, errorCode(error))
        } else {
            finalizeFailedBatch(claimed, errorCode(error))
        }
    }
}

private fun Services.claimOutboxBatch(limit: Int): ClaimedBatch? {
    val leaseToken = UUID.randomUUID().toString()
    val moment = OffsetDateTime.now(ZoneOffset.UTC)
    val leaseUntil = moment.plusNanos(OUTBOX_LEASE_MS * 1_000_000)

    return jdbi.tx { h ->
        h.deleteOrphanOutboxEntries()
        val retryable = h.selectRetryableBatchId(moment)
        val id: String
        if (retryable == null) {
            val entries = h.selectUnbatchedMessageIds(limit)
            if (entries.isEmpty()) return@tx null
            id = h.createOutboxBatch(leaseToken, leaseUntil, moment)
            h.assignEntriesToBatch(id, entries)
        } else {
            id = retryable
            h.reclaimBatch(id, leaseToken, leaseUntil)
        }
        val batch = h.findOutboxBatch(id)!!
        val messages = h.findBatchMessagesOrdered(id)
        if (batch.firstAttemptAt == null || messages.any { it.state != EmailMessageState.PENDING }) {
            throw IllegalStateException("Claimed outbox batch contains invalid message state")
        }
        ClaimedBatch(id, leaseToken, batch.attemptCount, batch.firstAttemptAt, messages)
    }
}

private fun Services.finalizeAcceptedBatch(batch: ClaimedBatch, resendEmailIds: List<String>): JsonObject {
    jdbi.tx { h ->
        assertLeaseOwner(h, batch)
        val updated = h.acceptBatchMessages(batch.messages.map { it.id }, resendEmailIds)
        if (updated != batch.messages.size) throw IllegalStateException("Outbox message changed before batch completion")
        resendEmailIds.forEach { projectOutboundDeliveryState(h, it) }
        if (h.deleteOwnedBatch(batch.id, batch.leaseToken) != 1) {
            throw IllegalStateException("Outbox batch lease was lost before cleanup")
        }
    }
    return drainResult(
        batchId = batch.id,
        claimed = batch.messages.size,
        accepted = batch.messages.size,
        results = batch.messages.mapIndexed { index, message ->
            DrainEntry(message.id, "accepted", resendEmailIds[index])
        },
    )
}

private fun Services.finalizeFailedBatch(batch: ClaimedBatch, errorCode: String): JsonObject {
    jdbi.tx { h ->
        assertLeaseOwner(h, batch)
        h.markBatchMessagesState(batch.messages.map { it.id }, EmailMessageState.FAILED, "Resend batch request failed ($errorCode)")
        if (h.deleteOwnedBatch(batch.id, batch.leaseToken) != 1) {
            throw IllegalStateException("Outbox batch lease was lost before cleanup")
        }
    }
    return terminalDrainResult(batch, "failed")
}

private fun Services.finalizeIndeterminateBatch(batch: ClaimedBatch, detail: String): JsonObject {
    jdbi.tx { h ->
        assertLeaseOwner(h, batch)
        h.markBatchMessagesState(batch.messages.map { it.id }, EmailMessageState.INDETERMINATE, detail)
        if (h.deleteOwnedBatch(batch.id, batch.leaseToken) != 1) {
            throw IllegalStateException("Outbox batch lease was lost before cleanup")
        }
    }
    return terminalDrainResult(batch, "indeterminate")
}

private fun Services.scheduleBatchRetryResult(batch: ClaimedBatch, errorCode: String): JsonObject {
    if (hasProviderWindowExpired(batch)) {
        return finalizeIndeterminateBatch(batch, "Outbox batch exceeded the provider idempotency window")
    }
    val delay = RETRY_DELAYS_MS[minOf(batch.attemptCount - 1, RETRY_DELAYS_MS.size - 1)]
    val nextAttemptAt = OffsetDateTime.now(ZoneOffset.UTC).plusNanos(delay * 1_000_000)
    val updated = jdbi.h { it.scheduleBatchRetry(batch.id, batch.leaseToken, nextAttemptAt, errorCode) }
    if (updated != 1) throw IllegalStateException("Outbox batch lease was lost before retry scheduling")
    return drainResult(
        batchId = batch.id,
        claimed = batch.messages.size,
        retryScheduled = batch.messages.size,
        results = batch.messages.map { DrainEntry(it.id, "pending", null) },
    )
}

private fun assertLeaseOwner(h: org.jdbi.v3.core.Handle, batch: ClaimedBatch) {
    if (!h.isBatchLeaseOwned(batch.id, batch.leaseToken)) {
        throw IllegalStateException("Outbox batch lease was lost before completion")
    }
}

private fun isRetryableBatchError(error: Throwable): Boolean {
    if (error !is ResendApiError) return true
    return (error.status == 429 && error.code != "monthly_quota_exceeded" && error.code != "daily_quota_exceeded") ||
        error.status >= 500 ||
        (error.status == 409 && error.code == "concurrent_idempotent_requests")
}

private fun hasProviderWindowExpired(batch: ClaimedBatch): Boolean =
    System.currentTimeMillis() - batch.firstAttemptAt.toInstant().toEpochMilli() >= PROVIDER_IDEMPOTENCY_SAFETY_MS

private fun errorCode(error: Throwable): String = when (error) {
    is ResendApiError -> error.code ?: "http_${error.status}"
    else -> error::class.simpleName ?: "unknown_error"
}

// --- result shaping ---

private data class DrainEntry(val messageId: String, val state: String, val resendEmailId: String?)

private fun drainResult(
    batchId: String?,
    claimed: Int = 0,
    accepted: Int = 0,
    failed: Int = 0,
    retryScheduled: Int = 0,
    indeterminate: Int = 0,
    results: List<DrainEntry> = emptyList(),
): JsonObject = buildJsonObject {
    put("batchId", batchId?.let { JsonPrimitive(it) } ?: JsonNull)
    put("claimed", JsonPrimitive(claimed))
    put("accepted", JsonPrimitive(accepted))
    put("failed", JsonPrimitive(failed))
    put("retryScheduled", JsonPrimitive(retryScheduled))
    put("indeterminate", JsonPrimitive(indeterminate))
    put("results", buildJsonArray {
        results.forEach { entry ->
            add(buildJsonObject {
                put("messageId", JsonPrimitive(entry.messageId))
                put("state", JsonPrimitive(entry.state))
                put("resendEmailId", entry.resendEmailId?.let { JsonPrimitive(it) } ?: JsonNull)
            })
        }
    })
}

private fun terminalDrainResult(batch: ClaimedBatch, state: String): JsonObject = drainResult(
    batchId = batch.id,
    claimed = batch.messages.size,
    failed = if (state == "failed") batch.messages.size else 0,
    indeterminate = if (state == "indeterminate") batch.messages.size else 0,
    results = batch.messages.map { DrainEntry(it.id, state, null) },
)

private fun emptyDrainResult(): JsonObject = drainResult(batchId = null)

File: src/lib/outbox-service.ts
import { randomUUID } from 'node:crypto';
import { type EmailMessage, Prisma, type PrismaClient } from '@/lib/database';
import {
  getConfiguredResendClient,
  ResendApiError,
  reconcileOutboundDeliveryState,
} from '@/lib/email';
import { buildSendEmailInput } from './conversation-service';

const OUTBOX_LEASE_MS = 2 * 60 * 1000;
const PROVIDER_IDEMPOTENCY_SAFETY_MS = 23 * 60 * 60 * 1000;
const RETRY_DELAYS_MS = [60_000, 120_000, 300_000] as const;

type BatchRow = { id: string };
type MessageIdRow = { message_id: string };

interface ClaimedBatch {
  id: string;
  leaseToken: string;
  attemptCount: number;
  firstAttemptAt: Date;
  messages: EmailMessage[];
}

export interface OutboxDrainResult {
  batchId: string | null;
  claimed: number;
  accepted: number;
  failed: number;
  retryScheduled: number;
  indeterminate: number;
  results: Array<{
    messageId: string;
    state: 'accepted' | 'failed' | 'pending' | 'indeterminate';
    resendEmailId: string | null;
  }>;
}

export async function drainEmailOutbox(
  client: PrismaClient,
  limit: number,
): Promise<OutboxDrainResult> {
  const claimed = await claimOutboxBatch(client, limit);
  if (!claimed) {
    return emptyDrainResult();
  }
  if (claimed.attemptCount > 1 && hasProviderWindowExpired(claimed)) {
    return finalizeIndeterminateBatch(
      client,
      claimed,
      'Outbox batch exceeded the provider idempotency window',
    );
  }

  try {
    const response = await getConfiguredResendClient().sendBatch(
      claimed.messages.map(buildSendEmailInput),
      `conversation-outbox/${claimed.id}`,
    );
    if (
      response.data.length !== claimed.messages.length ||
      response.data.some(
        (item) => typeof item.id !== 'string' || item.id.length === 0,
      )
    ) {
      return scheduleBatchRetry(client, claimed, 'invalid_batch_response');
    }
    return finalizeAcceptedBatch(
      client,
      claimed,
      response.data.map(({ id }) => id),
    );
  } catch (error) {
    if (
      error instanceof ResendApiError &&
      error.status === 409 &&
      error.code === 'invalid_idempotent_request'
    ) {
      return finalizeIndeterminateBatch(
        client,
        claimed,
        'Resend rejected a changed payload for the persisted batch key',
      );
    }
    if (isRetryableBatchError(error)) {
      return scheduleBatchRetry(client, claimed, getErrorCode(error));
    }
    return finalizeFailedBatch(client, claimed, getErrorCode(error));
  }
}

async function claimOutboxBatch(
  client: PrismaClient,
  limit: number,
): Promise<ClaimedBatch | null> {
  const leaseToken = randomUUID();
  const now = new Date();
  const leaseUntil = new Date(now.getTime() + OUTBOX_LEASE_MS);

  const batch = await client.$transaction(async (transaction) => {
    await transaction.$executeRaw`
      DELETE FROM email_outbox_entries AS entry
      USING email_messages AS message
      WHERE entry.message_id = message.id
        AND entry.batch_id IS NULL
        AND message.state <> 'PENDING'
    `;

    const retryable = await transaction.$queryRaw<BatchRow[]>`
      SELECT id
      FROM email_outbox_batches
      WHERE next_attempt_at <= ${now}
        AND (lease_until IS NULL OR lease_until <= ${now})
      ORDER BY next_attempt_at, id
      FOR UPDATE SKIP LOCKED
      LIMIT 1
    `;
    let batchId = retryable[0]?.id;

    if (!batchId) {
      const entries = await transaction.$queryRaw<MessageIdRow[]>`
        SELECT entry.message_id
        FROM email_outbox_entries AS entry
        INNER JOIN email_messages AS message ON message.id = entry.message_id
        WHERE entry.batch_id IS NULL
          AND message.state = 'PENDING'
        ORDER BY entry.queued_at, entry.message_id
        FOR UPDATE OF entry SKIP LOCKED
        LIMIT ${limit}
      `;
      if (!entries.length) {
        return null;
      }

      const created = await transaction.emailOutboxBatch.create({
        data: {
          leaseToken,
          leaseUntil,
          firstAttemptAt: now,
          attemptCount: 1,
        },
      });
      batchId = created.id;
      const assignments = Prisma.join(
        entries.map(
          (entry, position) =>
            Prisma.sql`(${entry.message_id}::uuid, ${position}::integer)`,
        ),
      );
      await transaction.$executeRaw`
        UPDATE email_outbox_entries AS entry
        SET batch_id = ${batchId}::uuid,
            batch_position = assignment.position
        FROM (VALUES ${assignments}) AS assignment(message_id, position)
        WHERE entry.message_id = assignment.message_id
          AND entry.batch_id IS NULL
      `;
    } else {
      await transaction.emailOutboxBatch.update({
        where: { id: batchId },
        data: {
          leaseToken,
          leaseUntil,
          attemptCount: { increment: 1 },
          lastErrorCode: null,
        },
      });
    }

    return transaction.emailOutboxBatch.findUniqueOrThrow({
      where: { id: batchId },
      include: {
        entries: {
          orderBy: { batchPosition: 'asc' },
          include: { message: true },
        },
      },
    });
  });

  if (!batch) {
    return null;
  }
  if (
    !batch.firstAttemptAt ||
    batch.entries.some((entry) => entry.message.state !== 'PENDING')
  ) {
    throw new Error('Claimed outbox batch contains invalid message state');
  }

  return {
    id: batch.id,
    leaseToken,
    attemptCount: batch.attemptCount,
    firstAttemptAt: batch.firstAttemptAt,
    messages: batch.entries.map((entry) => entry.message),
  };
}

async function finalizeAcceptedBatch(
  client: PrismaClient,
  batch: ClaimedBatch,
  resendEmailIds: string[],
): Promise<OutboxDrainResult> {
  await client.$transaction(async (transaction) => {
    await assertLeaseOwner(transaction, batch);
    const accepted = Prisma.join(
      batch.messages.map(
        (message, position) =>
          Prisma.sql`(${message.id}::uuid, ${resendEmailIds[position]}::text)`,
      ),
    );
    const updated = await transaction.$executeRaw`
      UPDATE email_messages AS message
      SET state = 'ACCEPTED',
          state_detail = NULL,
          delivery_state = 'UNKNOWN',
          resend_email_id = accepted.resend_email_id,
          updated_at = now()
      FROM (VALUES ${accepted}) AS accepted(message_id, resend_email_id)
      WHERE message.id = accepted.message_id
        AND message.state = 'PENDING'
    `;
    if (updated !== batch.messages.length) {
      throw new Error('Outbox message changed before batch completion');
    }
    for (const resendEmailId of resendEmailIds) {
      await reconcileOutboundDeliveryState(transaction, resendEmailId);
    }
    await deleteOwnedBatch(transaction, batch);
  });

  return {
    batchId: batch.id,
    claimed: batch.messages.length,
    accepted: batch.messages.length,
    failed: 0,
    retryScheduled: 0,
    indeterminate: 0,
    results: batch.messages.map((message, position) => ({
      messageId: message.id,
      state: 'accepted',
      resendEmailId: resendEmailIds[position],
    })),
  };
}

async function finalizeFailedBatch(
  client: PrismaClient,
  batch: ClaimedBatch,
  errorCode: string,
): Promise<OutboxDrainResult> {
  await client.$transaction(async (transaction) => {
    await assertLeaseOwner(transaction, batch);
    await transaction.emailMessage.updateMany({
      where: {
        id: { in: batch.messages.map(({ id }) => id) },
        state: 'PENDING',
      },
      data: {
        state: 'FAILED',
        stateDetail: `Resend batch request failed (${errorCode})`,
      },
    });
    await deleteOwnedBatch(transaction, batch);
  });

  return terminalDrainResult(batch, 'failed');
}

async function scheduleBatchRetry(
  client: PrismaClient,
  batch: ClaimedBatch,
  errorCode: string,
): Promise<OutboxDrainResult> {
  if (hasProviderWindowExpired(batch)) {
    return finalizeIndeterminateBatch(
      client,
      batch,
      'Outbox batch exceeded the provider idempotency window',
    );
  }

  const delay =
    RETRY_DELAYS_MS[
      Math.min(batch.attemptCount - 1, RETRY_DELAYS_MS.length - 1)
    ];
  const updated = await client.emailOutboxBatch.updateMany({
    where: { id: batch.id, leaseToken: batch.leaseToken },
    data: {
      nextAttemptAt: new Date(Date.now() + delay),
      leaseToken: null,
      leaseUntil: null,
      lastErrorCode: errorCode.slice(0, 64),
    },
  });
  if (updated.count !== 1) {
    throw new Error('Outbox batch lease was lost before retry scheduling');
  }

  return {
    batchId: batch.id,
    claimed: batch.messages.length,
    accepted: 0,
    failed: 0,
    retryScheduled: batch.messages.length,
    indeterminate: 0,
    results: batch.messages.map((message) => ({
      messageId: message.id,
      state: 'pending',
      resendEmailId: null,
    })),
  };
}

async function finalizeIndeterminateBatch(
  client: PrismaClient,
  batch: ClaimedBatch,
  detail: string,
): Promise<OutboxDrainResult> {
  await client.$transaction(async (transaction) => {
    await assertLeaseOwner(transaction, batch);
    await transaction.emailMessage.updateMany({
      where: {
        id: { in: batch.messages.map(({ id }) => id) },
        state: 'PENDING',
      },
      data: { state: 'INDETERMINATE', stateDetail: detail },
    });
    await deleteOwnedBatch(transaction, batch);
  });
  return terminalDrainResult(batch, 'indeterminate');
}

function terminalDrainResult(
  batch: ClaimedBatch,
  state: 'failed' | 'indeterminate',
): OutboxDrainResult {
  return {
    batchId: batch.id,
    claimed: batch.messages.length,
    accepted: 0,
    failed: state === 'failed' ? batch.messages.length : 0,
    retryScheduled: 0,
    indeterminate: state === 'indeterminate' ? batch.messages.length : 0,
    results: batch.messages.map((message) => ({
      messageId: message.id,
      state,
      resendEmailId: null,
    })),
  };
}

async function assertLeaseOwner(
  transaction: Prisma.TransactionClient,
  batch: ClaimedBatch,
) {
  const owned = await transaction.emailOutboxBatch.findFirst({
    where: { id: batch.id, leaseToken: batch.leaseToken },
    select: { id: true },
  });
  if (!owned) {
    throw new Error('Outbox batch lease was lost before completion');
  }
}

async function deleteOwnedBatch(
  transaction: Prisma.TransactionClient,
  batch: ClaimedBatch,
) {
  const deleted = await transaction.emailOutboxBatch.deleteMany({
    where: { id: batch.id, leaseToken: batch.leaseToken },
  });
  if (deleted.count !== 1) {
    throw new Error('Outbox batch lease was lost before cleanup');
  }
}

function isRetryableBatchError(error: unknown): boolean {
  if (!(error instanceof ResendApiError)) {
    return true;
  }
  return (
    (error.status === 429 &&
      error.code !== 'monthly_quota_exceeded' &&
      error.code !== 'daily_quota_exceeded') ||
    error.status >= 500 ||
    (error.status === 409 && error.code === 'concurrent_idempotent_requests')
  );
}

function hasProviderWindowExpired(batch: ClaimedBatch): boolean {
  return (
    Date.now() - batch.firstAttemptAt.getTime() >=
    PROVIDER_IDEMPOTENCY_SAFETY_MS
  );
}

function getErrorCode(error: unknown): string {
  if (error instanceof ResendApiError) {
    return error.code ?? `http_${error.status}`;
  }
  return error instanceof Error ? error.name : 'unknown_error';
}

function emptyDrainResult(): OutboxDrainResult {
  return {
    batchId: null,
    claimed: 0,
    accepted: 0,
    failed: 0,
    retryScheduled: 0,
    indeterminate: 0,
    results: [],
  };
}
